crate/cli/src/actions/google/keypairs/insert_keypairs.rs

```diff
@@ -0,0 +1,169 @@
+use std::{
+    collections::HashMap,
+    fs::{self, File},
+    io::Read,
+    path::PathBuf,
+};
+
+use clap::Parser;
+use serde::{Deserialize, Serialize};
+
+use super::KEYPAIRS_ENDPOINT;
+use crate::{actions::google::gmail_client::GmailClient, error::CliError};
+
+/// Creates and uploads a client-side encryption S/MIME public key certificate chain and private key metadata for a user.
+#[derive(Parser)]
+#[clap(verbatim_doc_comment)]
+pub struct InsertKeypairsAction {
+    /// The requester's primary email address
+    #[clap(long = "user-id", short = 'u', required = true)]
+    user_id: String,
+
+    /// Input directory with wrapped key files, with email as basename
+    #[clap(long = "inkeydir", short = 'k', required = true)]
+    inkeydir: PathBuf,
+
+    /// Input directory with p7 pem certs with extension p7pem, with email as basename
+    #[clap(long = "incertdir", short = 'c', required = true)]
+    incertdir: PathBuf,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct KeyFile {
+    kacls_url: String,
+    wrapped_private_key: String,
+}
+
+#[derive(Serialize, Deserialize)]
+#[allow(non_snake_case)]
+struct KeyPairInfo {
+    pkcs7: String,
+    privateKeyMetadata: Vec<PrivateKeyMetadata>,
+}
+
+#[derive(Serialize, Deserialize)]
+#[allow(non_snake_case)]
+struct PrivateKeyMetadata {
+    kaclsKeyMetadata: KaclsKeyMetadata,
+}
+
+#[derive(Serialize, Deserialize)]
+#[allow(non_snake_case)]
+struct KaclsKeyMetadata {
+    kaclsUri: String,
+    kaclsData: String,
+}
+
+impl InsertKeypairsAction {
+    fn get_input_files(indir: &PathBuf, ext: &str) -> Result<Vec<PathBuf>, CliError> {
+        let full_names: Vec<PathBuf> = fs::read_dir(indir)?
+            .filter_map(|entry| entry.ok().map(|e| e.path()))
+            .collect();
+
+        let all_files: Vec<PathBuf> = full_names.iter().filter(|f| f.is_file()).cloned().collect();
+
+        let input_files: Vec<PathBuf> = all_files
+            .into_iter()
+            .filter(|f| f.extension().map_or(false, |e| e == ext))
+            .collect();
+
+        Ok(input_files)
+    }
+
+    fn get_email_to_file(
+        files: &[PathBuf],
+        ext: &str,
+    ) -> Result<HashMap<String, PathBuf>, CliError> {
+        let mut email_file_map = HashMap::new();
+
+        for file in files {
+            let fname = file
+                .file_name()
+                .ok_or_else(|| {
+                    CliError::Conversion(format!("cannot get file name from input file {file:?}",))
+                })?
+                .to_string_lossy();
+            let (email, xtn) = match fname.rfind('.') {
+                Some(idx) => fname.split_at(idx),
+                None => continue,
+            };
+
+            if xtn[1..] != *ext {
```

Comment:
This code will panic is `xtn` is `&""`. This may not happen, but deserves a comment.